sys/vane/alef.hoon

```diff
@@ -0,0 +1,2588 @@
+::    requested the original "forward flow" message be sent.
+::
+::    A "backward flow" message, which is similar to a response or a
+::    subscription update, is given to Ames from a local vane.  Ames
+::    transmits the message to the peer's Ames, which gives the message
+::    to the destination vane.
+::
+::    Ames will give a %memo to a vane upon hearing the message from a
+::    remote. This message is a "backward flow" message, forming one of
+::    potentially many responses to a "forward flow" message that a
+::    local vane had passed to our local Ames, and which local Ames had
+::    relayed to the remote.  Ames gives the %memo on the same duct the
+::    local vane had originally used to pass Ames the "forward flow"
+::    message.
+::
+::    Backward flow messages are acked automatically by the receiver.
+::    They cannot be nacked, and Ames only uses the ack internally,
+::    without notifying the client vane.
+::
+::    Forward flow messages can be nacked, in which case the peer will
+::    send both a message-nack packet and a nack-trace message, which is
+::    sent on a special diagnostic flow so as not to interfere with
+::    normal operation.  The nack-trace is sent as a full Ames message,
+::    instead of just a packet, because the contained error information
+::    can be arbitrarily large.
+::
+::    Once the local Ames has received the nack-trace, it knows the peer
+::    has received the full message and failed to process it.  This
+::    means if we later hear an ack packet on the failed message, we can
+::    ignore it.
+::
+::    Also, due to Ames's exactly-once delivery semantics, we know that
+::    when we receive a nack-trace for message n, we know the peer has
+::    positively acked all messages m+1 through n-1, where m is the last
+::    message for which we heard a nack-trace.  If we haven't heard acks
+::    on all those messages, we apply positive acks when we hear the
+::    nack-trace.
+::
+::    %memo: message to vane from peer
+::    %send: packet to unix
+::    %done: notify vane that peer (n)acked our message
+::
+::    %mass: memory usage report
+::    %turf: domain report, relayed from jael
+::
++$  gift
+  $%  [%memo =message]
+      [%send =lane =blob]
+      [%done error=(unit error)]
+  ::
+      [%mass mass]
+      [%turf turfs=(list turf)]
+  ==
+::  $note: request to other vane
+::
+::    TODO: specialize gall interface for subscription management
+::
+::    Ames passes a %memo note to another vane when it receives a
+::    message on a "forward flow" from a peer, originally passed from
+::    one of the peer's vanes to the peer's Ames.
+::
+::    Ames passes a %memo to itself to trigger a heartbeat message to
+::    our sponsor.
+::
+::    Ames passes a %private-keys to Jael to request our private keys.
+::    Ames passes a %public-keys to Jael to request a peer's public
+::    keys.
+::
++$  note
+  $%  $:  %a
+      $%  [%memo sponsor=ship message=_[/a/ping ~]]
```

Comment:
I think these `+note` types are unnecessarily over-constrained (some with `+sign` below). The interface ames exposes generalizes without any change to every vane in the system. I'd be inclined to prefix `%memo` here (and `%done` in `+sign`) with just `@tas`.

Reply:
Yeah, I think you're right